### PR TITLE
Make `chop` and `chomp` return `SubString`

### PR DESCRIPTION
**Pull request by TotalVerb**
This is unfortunate:

``` julia
julia> const str = "x" ^ 100000 * "\n";

julia> @benchmark split(str, '\n')[1]
BenchmarkTools.Trial: 
  samples:          10000
  evals/sample:     8
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  192.00 bytes
  allocs estimate:  4
  minimum time:     2.99 μs (0.00% GC)
  median time:      3.21 μs (0.00% GC)
  mean time:        3.26 μs (0.00% GC)
  maximum time:     8.44 μs (0.00% GC)

julia> @benchmark chomp(str)
BenchmarkTools.Trial: 
  samples:          10000
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  97.78 kb
  allocs estimate:  3
  minimum time:     9.35 μs (0.00% GC)
  median time:      10.81 μs (0.00% GC)
  mean time:        13.47 μs (10.56% GC)
  maximum time:     790.80 μs (94.19% GC)
```

I suspect that the same reason that `split` returns `SubString` applies to `chop` and `chomp` also. The clear benefit of returning a view is performance and allocation avoidance; the disadvantages are that the underlying string cannot be freed, and accessing a `SubString` is slightly slower. However, in `chop` and `chomp`'s case, the disadvantage is incredibly slight—an extra byte or so that can't be freed—so I believe it would make sense to follow `split` here.

Note that Perl, the only other language I know with `chop` and `chomp`, has these as _in-place_ operations—which is even faster than creating a `SubString`, but not appropriate for Julia's conventionally immutable `String`s.

On another note, I wonder whether it might be worthwhile to make `String` into `String{T}`, so that a string's underlying data can be an array view—which would make `SubString` substantially faster.

### CODE REVIEW

**Comment by nalimilan** [comment]
I think @StefanKarpinski has a plan regarding `SubString` and strings sharing the same underlying array.

**Comment by stevengj** [comment]
LGTM (needs a rebase).

**Comment by TotalVerb** [comment]
Rebased. On another note, I noticed the doctest:

```
julia> a = string("March")
"March"
julia> chop(a)
"Marc"
```

seems to use `string("March")` for no apparent reason. I think this is a little misleading since it seems to imply copying the string is needed to use `chop`.

**Comment by tkelman** [comment]
The return type should be tested for if you don't want it to regress accidentally. The `string` call is probably a bad find-replace from `utf8`.

**Comment by stevengj** [comment]
I would just fix the doctest in this PR, @TotalVerb.

**Comment by StefanKarpinski** [comment]
LGTM as well.

**Comment by TotalVerb** [comment]
I made the doctest fix locally. Do I need to rebuild helpdb or something? @stevengj

**Comment by stevengj** [comment]
You change the docstring [in the source](https://github.com/JuliaLang/julia/blob/b1f152597b5c263b0e75acec6661a6b48332b896/base/strings/util.jl#L68), run `./julia doc/genstdlib.jl` to re-generate the manual, and then commit the changes.

**Comment by TotalVerb** [comment]
OK, is this all right? Thanks for the help, @stevengj.

**Comment by stevengj** [comment]
LGTM.

File: base/strings/util.jl [hunk 1]
```diff
@@ -58,21 +58,20 @@ startswith(a::Vector{UInt8}, b::Vector{UInt8}) =
...
-julia> a = string("March")
+julia> a = "March"
 "March"
 
 julia> chop(a)
 "Marc"
 ```
 """
-chop(s::AbstractString) = s[1:end-1]
+chop(s::AbstractString) = SubString(s, 1, endof(s)-1)
```
Comment (kshyatt) [inline base/strings/util.jl]: Do the doctests still pass with this change?

Comment (TotalVerb) [inline base/strings/util.jl]: The doctests don't pass, but I don't know if that's this change's fault. Failures seem unrelated to me.

Comment (TotalVerb) [inline base/strings/util.jl]: It looks to me like most of the failures are due to line numbers being not the same as expected?

https://gist.github.com/TotalVerb/d81e7fff617d315f51ecd019aac4f553

**Comment by TotalVerb** [comment]
Bump.

**Comment by KristofferC** [comment]
This got lgtm from both Stefan and Steven so should be OK to merge?

**Review by kmsquire — COMMENTED** [review]
News item?

**Comment by kmsquire** [comment]
Sorry, that should have been a comment.

**Comment by TotalVerb** [comment]
Sure, I'll write up a NEWS.

**Pull request merged**